src/include/cache.i

```diff
@@ -485,6 +485,13 @@ __wt_cache_eviction_check(WT_SESSION_IMPL *session, bool busy, bool readonly, bo
             WT_SESSION_LOCKED_TABLE))
         return (0);
 
+    /*
+     * We can re-enter eviction code when we are trying to start a new transaction for eviction
+     * thread workers.
+     */
+    if (F_ISSET(session, WT_SESSION_INTERNAL) && !WT_SESSION_IS_CHECKPOINT(session))
```

Comment:
Does this mean that other internal operations will skip helping with eviction, when maybe they shouldn't? i.e: for tests like this I'd prefer they are as narrow as possible.

Reply:
I believe I can make this check more strict to make sure we are only returning early when running in context of eviction threads

Reply:
I have added a new flag to indicate that the session belongs to dedicated eviction thread in 8076bb9.